encoding/ccf/decode.go

```diff
@@ -0,0 +1,1887 @@
+		)
+	}
+	return cadence.NewMeteredWord8(d.gauge, uint8(i)), nil
+}
+
+// decodeWord16 decodes word16-value as
+// language=CDDL
+// word16-value = uint .le 65535
+func (d *Decoder) decodeWord16() (cadence.Value, error) {
+	i, err := d.dec.DecodeUint64()
+	if err != nil {
+		return nil, err
+	}
+	if i > math.MaxUint16 {
+		return nil, fmt.Errorf(
+			"encoded word16-value %d is outside range of Word16 [0, %d]",
+			i,
+			math.MaxUint16,
+		)
+	}
+	return cadence.NewMeteredWord16(d.gauge, uint16(i)), nil
+}
+
+// decodeWord32 decodes word32-value as
+// language=CDDL
+// word32-value = uint .le 4294967295
+func (d *Decoder) decodeWord32() (cadence.Value, error) {
+	i, err := d.dec.DecodeUint64()
+	if err != nil {
+		return nil, err
+	}
+	if i > math.MaxUint32 {
+		return nil, fmt.Errorf(
+			"encoded word32-value %d is outside range of Word32 [0, %d]",
+			i,
+			math.MaxUint32,
+		)
+	}
+	return cadence.NewMeteredWord32(d.gauge, uint32(i)), nil
+}
+
+// decodeWord64 decodes word64-value as
+// language=CDDL
+// word64-value = uint .le 18446744073709551615
+func (d *Decoder) decodeWord64() (cadence.Value, error) {
+	i, err := d.dec.DecodeUint64()
+	if err != nil {
+		return nil, err
+	}
+	return cadence.NewMeteredWord64(d.gauge, i), nil
+}
+
+// decodeFix64 decodes fix64-value as
+// language=CDDL
+// fix64-value = (int .ge -9223372036854775808) .le 9223372036854775807
+func (d *Decoder) decodeFix64() (cadence.Value, error) {
+	i, err := d.dec.DecodeInt64()
+	if err != nil {
+		return nil, err
+	}
+	return cadence.NewMeteredFix64FromInt64(d.gauge, i)
+}
+
+// decodeUFix64 decodes ufix64-value as
+// language=CDDL
+// ufix64-value = uint .le 18446744073709551615
+func (d *Decoder) decodeUFix64() (cadence.Value, error) {
+	i, err := d.dec.DecodeUint64()
+	if err != nil {
+		return nil, err
+	}
+	return cadence.NewMeteredUFix64FromUint64(d.gauge, i)
+}
+
+// decodeOptional decodes encoded optional-value as
+// language=CDDL
+// optional-value = nil / value
+func (d *Decoder) decodeOptional(typ *cadence.OptionalType, types cadenceTypeByCCFTypeID) (cadence.Value, error) {
+	// Peek ahead for next CBOR data item type
+	nextType, err := d.dec.NextType()
+	if err != nil {
+		return nil, err
+	}
+
+	switch nextType {
+	case cbor.NilType:
+		// Decode nil.
+		err := d.dec.DecodeNil()
+		if err != nil {
+			return nil, err
+		}
+		return newNilOptionalValue(d.gauge, typ), nil
+
+	default:
+		// Decode value.
+		value, err := d.decodeValue(typ.Type, types)
+		if err != nil {
+			return nil, err
+		}
+		return cadence.NewMeteredOptional(d.gauge, value), nil
+	}
+}
+
+// decodeArray decodes encoded array-value as
+// language=CDDL
+// array-value = [* value]
+func (d *Decoder) decodeArray(typ cadence.ArrayType, hasKnownSize bool, knownSize uint, types cadenceTypeByCCFTypeID) (cadence.Value, error) {
+	// Decode array length.
+	n, err := d.dec.DecodeArrayHead()
+	if err != nil {
+		return nil, err
+	}
+
+	if hasKnownSize && uint64(knownSize) != n {
```

Comment:
Maybe just change the parameter to `knownSize uint64`

```suggestion
	if hasKnownSize && knownSize != n {
```

Reply:
Done in commit 8791232f56fcd97771f3d9a255b524e73624fc9f.